File: java/src/main/java/com/android/util/SizeUtil.java
package com.android.util;

import com.android.base.BaseApplication;

/**
 * Created by xuzhb on 2019/10/19
 * Desc:尺寸信息单位转换
 */
public class SizeUtil {

    public static float px2dp(float pxValue) {
        float density = BaseApplication.getInstance().getResources().getDisplayMetrics().density;
        return pxValue / density + 0.5f;
    }

    public static float dp2px(float dpValue) {
        float density = BaseApplication.getInstance().getResources().getDisplayMetrics().density;
        return dpValue * density + 0.5f;
    }

    public static float px2sp(float pxValue) {
        float scaledDensity = BaseApplication.getInstance().getResources().getDisplayMetrics().scaledDensity;
        return pxValue / scaledDensity + 0.5f;
    }

    public static float sp2px(float spValue) {
        float scaledDensity = BaseApplication.getInstance().getResources().getDisplayMetrics().scaledDensity;
        return spValue * scaledDensity + 0.5f;
    }

    public static int px2dpInt(float pxValue) {
        return (int) px2dp(pxValue);
    }

    public static int dp2pxInt(float dpValue) {
        return (int) dp2px(dpValue);
    }

    public static int px2spInt(float pxValue) {
        return (int) px2sp(pxValue);
    }

    public static int sp2pxInt(float spValue) {
        return (int) sp2px(spValue);
    }

}

File: kotlin/src/main/java/com/android/util/SizeUtil.kt
package com.android.util

import android.content.Context
import com.android.base.BaseApplication

/**
 * Created by xuzhb on 2019/8/1
 * Desc:尺寸信息单位转换
 */
object SizeUtil {

    fun px2dp(pxValue: Float, context: Context = BaseApplication.instance): Float {
        val density: Float = context.resources.displayMetrics.density
        return pxValue / density + 0.5f
    }

    fun dp2px(dpValue: Float, context: Context = BaseApplication.instance): Float {
        val density: Float = context.resources.displayMetrics.density
        return dpValue * density + 0.5f
    }

    fun px2sp(pxValue: Float, context: Context = BaseApplication.instance): Float {
        val scaledDensity: Float = context.resources.displayMetrics.scaledDensity
        return pxValue / scaledDensity + 0.5f
    }

    fun sp2px(spValue: Float, context: Context = BaseApplication.instance): Float {
        val scaledDensity: Float = context.resources.displayMetrics.scaledDensity
        return spValue * scaledDensity + 0.5f
    }

    fun px2dpInt(pxValue: Float, context: Context = BaseApplication.instance): Int {
        return px2dp(pxValue, context).toInt()
    }

    fun dp2pxInt(dpValue: Float, context: Context = BaseApplication.instance): Int {
        return dp2px(dpValue, context).toInt()
    }

    fun px2spInt(pxValue: Float, context: Context = BaseApplication.instance): Int {
        return px2sp(pxValue, context).toInt()
    }

    fun sp2pxInt(spValue: Float, context: Context = BaseApplication.instance): Int {
        return sp2px(spValue, context).toInt()
    }

}